### PR TITLE
feat: add vision 2D RoPE ref, npu ttx impl and tests.

### PR DESCRIPTION
**Pull request by Xu-pan**

### CODE REVIEW

File: mojo_opset/tests/accuracy/operators/_vision_rope_2d_golden.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+import torch
```
Comment (wwens7) [inline mojo_opset/tests/accuracy/operators/_vision_rope_2d_golden.py]: 没必要单抽一个文件出来吧，而且也不需要对比 torch vs torch 的精度？

File: mojo_opset/tests/accuracy/operators/test_position_embedding.py [hunk 2]
```diff
@@ -1,16 +1,33 @@
...
 from mojo_opset import MojoApplyRoPE
 from mojo_opset import MojoGridRoPE
-from mojo_opset import MojoRelativeEmbedding
-from mojo_opset import MojoRotaryEmbedding
+from mojo_opset import MojoVisionRoPE2D
 from mojo_opset.utils.platform import get_torch_device
 
 torch.random.manual_seed(42)
 
+VISION_VIT_CONFIG = {
```
Comment (wwens7) [inline mojo_opset/tests/accuracy/operators/test_position_embedding.py:15]: 一组 config 够不？这里的一组是实际vit 的 config 里摘出来的吗

File: mojo_opset/tests/accuracy/operators/test_position_embedding.py [hunk 3]
```diff
@@ -209,27 +226,90 @@ def test_grid_pos_emb(bs, grid, heads, head_dim, pad, dtype):
...
+        pytest.param(VISION_VIT_CONFIG, id="vision_448_27l_20h_h64"),
     ],
 )
+@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16])
 @bypass_not_implemented
-def test_relative_embedding(num_buckets, num_heads, bidirectional, lq, lk):
-    emb = MojoRelativeEmbedding(num_buckets=num_buckets, num_heads=num_heads, bidirectional=bidirectional)
-    emb_ref = MojoRelativeEmbedding._registry.get("torch")(
-        num_buckets=num_buckets, num_heads=num_heads, bidirectional=bidirectional
+def test_vision_rope_2d_packed_against_golden(
```
Comment (wwens7) [inline mojo_opset/tests/accuracy/operators/test_position_embedding.py]: 这里参考其他测试写一个 torch 后端 vs 特定后端的测试吧，对比 golden 和其他边界情况的可以暂时不要？

File: mojo_opset/core/operators/position_embedding.py [hunk 4]
```diff
@@ -298,3 +225,121 @@ def forward(
...
+        rope_dim = head_dim if self.rope_dim is None else self.rope_dim
+        assert rope_dim == head_dim, "vision rope rotates the full head_dim"
+        assert rope_dim % 4 == 0, "vision 2D rope_dim must be divisible by 4"
+        total_tokens = int(grid_hw.to(dtype=torch.int64).prod(dim=-1).sum().item())
+        assert q.shape[0] == total_tokens, "packed token count must equal sum(grid_hw[:, 0] * grid_hw[:, 1])"
+        assert k.shape[0] == total_tokens, "packed token count must equal sum(grid_hw[:, 0] * grid_hw[:, 1])"
+
+        # The packed sequence shares one concatenated frequency table whose order
+        # follows the native multi-image patch packing order.
+        freqs = self._build_freqs(grid_hw, rope_dim, device=q.device)
```
Comment (wwens7) [inline mojo_opset/core/operators/position_embedding.py]: 这里_build_freqs 每次 fwd 都会重算一遍，可以参考普通 RoPE（分成了俩阶段，定义了俩 Mojo op）改造一下，这样一个 step 的所有层可以共享同个 freqs

File: mojo_opset/tests/accuracy/operators/_vision_rope_2d_golden.py (hunk 1 above)
Comment (Xu-pan) [reply mojo_opset/tests/accuracy/operators/_vision_rope_2d_golden.py]: 没错，是第一版证明ref和golden是一致的，正式版不带这些，pytest就直接backend vs ref

File: mojo_opset/tests/accuracy/operators/test_position_embedding.py (hunk 2 above)
Comment (Xu-pan) [reply mojo_opset/tests/accuracy/operators/test_position_embedding.py:15]: 是的，与实际config一致

File: mojo_opset/tests/accuracy/operators/test_position_embedding.py (hunk 3 above)
Comment (Xu-pan) [reply mojo_opset/tests/accuracy/operators/test_position_embedding.py]: 要的，其他后端有了就加上。我先看看torch npu有没有

File: mojo_opset/core/operators/position_embedding.py (hunk 4 above)
Comment (Xu-pan) [reply mojo_opset/core/operators/position_embedding.py]: 可以的，build 和 apply分开对吧？

File: mojo_opset/tests/accuracy/operators/test_position_embedding.py (hunk 3 above)
Comment (wwens7) [reply mojo_opset/tests/accuracy/operators/test_position_embedding.py]: 先加吧 这个合了之后我补个 triton 就可以测了

File: mojo_opset/core/operators/position_embedding.py (hunk 4 above)
Comment (wwens7) [reply mojo_opset/core/operators/position_embedding.py]: 对的

Comment (Xu-pan) [reply mojo_opset/core/operators/position_embedding.py]: 已分开，review哈

File: mojo_opset/tests/accuracy/operators/test_position_embedding.py [hunk 5]
```diff
@@ -233,3 +253,129 @@ def test_relative_embedding(num_buckets, num_heads, bidirectional, lq, lk):
...
+    adapooling_factor = vision_config["adapooling_factor"]
+
+    grid_hw = torch.tensor(grid, device=device, dtype=torch.int32)
+    rot_pos_emb = MojoVisionRotaryEmbedding2D(
+        rope_theta=rope_theta,
+        rope_dim=head_dim,
+        adapooling_factor=adapooling_factor,
+    ).to(device)
+
+    cos_ref, sin_ref = _vision_rotary_embedding_2d_golden(
```
Comment (wwens7) [inline mojo_opset/tests/accuracy/operators/test_position_embedding.py]: 参考其他测试实例化两个对象，一个是get("torch")来的 torch 实现，一个是直接实例化 MojoVisionRotaryEmbedding2D 的实现，加上bypass_not_impl 的装饰器，当前就会跳过测试了，后面我加上 ttx 也可以正常跑，_vision_rope_2d_golden.py也可以删掉了

Comment (Xu-pan) [reply mojo_opset/tests/accuracy/operators/test_position_embedding.py]: ok已经清理了，刚提了个torch npu的backend。看下哈

File: mojo_opset/tests/accuracy/operators/test_position_embedding.py (hunk 2 above)
Comment (Xu-pan) [reply mojo_opset/tests/accuracy/operators/test_position_embedding.py:15]: 需要补更多吗？之前构造非实际case有啥原则、规律么

File: mojo_opset/backends/ttx/kernels/npu/vision_rope.py [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+    seq = torch.arange(max_grid_size, device=device, dtype=torch.float32)
+    rotary_pos_emb_full = torch.outer(seq, inv_freq.to(device=device))
+
+    pos_ids = _build_position_ids(grid_hw, adapooling_factor, device=device)
+    freqs = rotary_pos_emb_full[pos_ids].flatten(-2)
+    emb = torch.cat([freqs, freqs], dim=-1)
+    return emb.cos(), emb.sin()
+
+
+def _build_position_ids(
```
Comment (zhangjihang-BD) [inline mojo_opset/backends/ttx/kernels/npu/vision_rope.py:49]: 这个函数感觉极其graph不友好，有办法device做么

File: mojo_opset/backends/torch_npu/operators/position_embedding.py [hunk 7]
```diff
@@ -50,3 +51,23 @@ def _apply_rope(
...
+        sin: torch.Tensor,
+    ) -> torch.Tensor:
+        orig_dtype = x.dtype
+
+        # npu_rotary_mul requires 4D input.
+        x = x.unsqueeze(0)
+        cos = cos.unsqueeze(0).unsqueeze(2)
+        sin = sin.unsqueeze(0).unsqueeze(2)
+
+        output = torch_npu.npu_rotary_mul(x, cos, sin, rotary_mode="half")
```
Comment (zhangjihang-BD) [inline mojo_opset/backends/torch_npu/operators/position_embedding.py:72]: 诶，为啥这里可以直接用torch_npu.npu_rotary_mul

**Review by zhangjihang-BD — APPROVED** [review]
